PR: Refactor repository structure.

davisadam10 opened the pull request:
DO NOT MERGE

Based on the last meeting I have taken onboard the feedback and made the changes

1) New Branch (this one without the git lfs)
2) Moved all settings/params into the projectSettings - (few remaining)
3) Generators now moved over to use abstract methods
4) Application & Generators totally decoupled, generators only refer to a project settings object
5) pre-commit hooks setup and running etc


Question

Some of the params which are set, such as reference colour checker, optimization factory are actually large arrays of data, or functions. I don't think we should be serialising down these objects, moreover we should probably store just the string values for these types of options, and the use either a factory or helper function to get the required object based on the stored keys.

If thats generally good, then i can look to finish the last few params and then its all done pending the ui updates

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/ampas/idt-calculator?pullRequest=87) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/ampas/idt-calculator?pullRequest=87) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/ampas/idt-calculator?pullRequest=87) it.</sub>

davisadam10 commented:
As discussed on the call, we will store the colour version within the spec and then store any colour objects or functions as string representations, and retrieve them from a factory class vs serializing the objects them selves.

When loading we will warn that the colour versions do not match if. they are different

@KelSolaar

davisadam10 commented:
Finished moving all settings to project settings.
Still do not merge, 
Needs review and updates to the UI to use the IDTApplication instead of the generator
Needs logic from ui moving down in to the IDTApplication

@KelSolaar know you've been travelling but if you get a chance to look at this would be great

KelSolaar commented on aces/idt/core/common.py:
```diff
@@ -47,14 +54,47 @@
 
-SDS_COLORCHECKER_CLASSIC = tuple(SDS_COLOURCHECKERS["ISO 17321-1"].values())
+
+def get_sds_colour_checker(colour_checker_name):
+    """
+    Get the Reference reflectances for the given colour checker
+
+    Parameters
+    ----------
+    colour_checker_name: The name of the colour checker name
+
+    Returns
+    -------
+        SDS_COLORCHECKER : tuple
+
+    """
+    return tuple(SDS_COLOURCHECKERS[colour_checker_name].values())
+
+
+def get_sds_illuminant(illuminant_name):
+    """
+    Get *ACES* reference illuminant spectral distribution, for the given illuminant name
+
+    Parameters
+    ----------
+    illuminant_name: The name of the illuminant
+
+    Returns
+    -------
+        SpectralDistribution
+
+    """
+    return SDS_ILLUMINANTS[illuminant_name]
```
**Minor**

Isn't that a bit "busy" for what a very simple lookups? They don't seem to be used a lot either.

davisadam10 replied:
So looking at this, the only reason these seem to exist is because in common.py we have these constants

SDS_COLORCHECKER_CLASSIC
RGB_COLORCHECKER_CLASSIC_ACES
SD_ILLUMINANT_ACES

These are not used anywhere else in the project and the

generate_reference_colour_checker takes them as default options on the input.

If we remove the default options to generate_reference_colour_checker

the values are already defaulted in the projectSettings, so we could then just loose all of this and move just the

SDS_ILLUMINANTS[illuminant_name] & tuple(SDS_COLOURCHECKERS[colour_checker_name].values())
into the project settings which it then passes into the generate_reference_colour_checker.

This removes 3 unused constants, and 2 un needed functions into a single function within the project_settings

KelSolaar commented on aces/idt/core/constants.py:
```diff
@@ -0,0 +1,412 @@
+    @classmethod
+    @property
+    def ALL(cls):
+        """
+
+        Returns A list of all the CATs
+        -------
```
**Pedantry**

We are not building docs atm but pretty sure that Sphinx would be sad here.

davisadam10 replied:
Done

KelSolaar commented on aces/idt/core/constants.py:
```diff
@@ -0,0 +1,412 @@
+        -------
+
+        """
+        cats = list(colour.CHROMATIC_ADAPTATION_TRANSFORMS.keys())
+        cats.sort()
+        return cats
```
**Minor**

Prefer this:

```python
>>> import colour
>>> sorted(colour.CHROMATIC_ADAPTATION_TRANSFORMS)
['Bianco 2010', 'Bianco PC 2010', 'Bradford', 'CAT02', 'CAT02 Brill 2008', 'CAT16', 'CMCCAT2000', 'CMCCAT97', 'Fairchild', 'Sharp', 'Von Kries', 'XYZ Scaling']
```

davisadam10 replied:
Done thanks knew there was a way

KelSolaar commented on aces/idt/core/constants.py:
```diff
@@ -0,0 +1,412 @@
+        default_value="ISO 17321-1",
+        description="The reference colour checker we want to use",
+        ui_type=UITypes.OPTIONS_FIELD,
+        options=["ISO 17321-1"],
```
```python
>>> sorted(colour.SDS_COLOURCHECKERS)
['BabelColor Average', 'ColorChecker N Ohta', 'ISO 17321-1', 'babel_average', 'cc_ohta']
```

davisadam10 replied:
Thanks done

KelSolaar commented on aces/idt/core/constants.py:
```diff
@@ -0,0 +1,412 @@
+        default_value="D60",
+        description="The illuminant we want to use for the reference colour checker",
+        ui_type=UITypes.OPTIONS_FIELD,
+        options=["D60"],
```
```python
>>> sorted(colour.SDS_ILLUMINANTS)
['A', 'B', 'C', 'D50', 'D55', 'D60', 'D65', 'D75', 'E', 'FL1', 'FL10', 'FL11', 'FL12', 'FL2', 'FL3', 'FL3.1', 'FL3.10', 'FL3.11', 'FL3.12', 'FL3.13', 'FL3.14', 'FL3.15', 'FL3.2', 'FL3.3', 'FL3.4', 'FL3.5', 'FL3.6', 'FL3.7', 'FL3.8', 'FL3.9', 'FL4', 'FL5', 'FL6', 'FL7', 'FL8', 'FL9', 'HP1', 'HP2', 'HP3', 'HP4', 'HP5', 'ID50', 'ID65', 'ISO 7589 Photoflood', 'ISO 7589 Photographic Daylight', 'ISO 7589 Sensitometric Daylight', 'ISO 7589 Sensitometric Photoflood', 'ISO 7589 Sensitometric Printer', 'ISO 7589 Sensitometric Studio Tungsten', 'ISO 7589 Studio Tungsten', 'LED-B1', 'LED-B2', 'LED-B3', 'LED-B4', 'LED-B5', 'LED-BH1', 'LED-RGB1', 'LED-V1', 'LED-V2']
```

davisadam10 replied:
Thanks Done

KelSolaar commented on aces/idt/core/structures.py:
```diff
@@ -0,0 +1,197 @@
+"""Module contains a selection of helper classes used throughout the IDT
```
**Pedantry**

Maybe something like this?

> Define various of helper classes used throughout the IDT generator.

KelSolaar commented on aces/idt/core/structures.py:
```diff
@@ -0,0 +1,197 @@
+    DATA = "data"
+
+
+class IDTMetaData:
```
**Major**

Seems like this wants to be a `dataclass`.

davisadam10 replied:
Are you referring to the IDTMetaData? or the SerializableConstants? or both?

Not used dataclass much before whats the benefit?

The SerializableConstants is essentially an Enum which probably should use the Enum python type

davisadam10 replied:
Resolved By Thomas

KelSolaar commented on aces/idt/core/structures.py:
```diff
@@ -0,0 +1,197 @@
+
+        Parameters
+        ----------
+        obj : object we want to serialize
```
**Minor**

We prefer Numpy style docstrings:

```
        Parameters
        ----------
        obj : object
            Object we want to serialize.
```

davisadam10 replied:
done will keep an eye out

KelSolaar commented on aces/idt/generators/base_generator.py:
```diff
@@ -0,0 +1,745 @@
+        # Disabling orientation as we now have an oriented quadrilateral
+        settings.reference_values = None
+
+        # TODO Again are we using this?
+        # Flatfield
```
For v2 yeah! :)

davisadam10 replied:
Ok so lets include flatfield through out while we are writing the cleanup, I didnt know if this was old, new, or proposed lol :)

davisadam10 replied:
Ok removed TODOs and matched as a constant like the other folders

KelSolaar commented on aces/idt/generators/prosumer_camera.py:
```diff
@@ -0,0 +1,602 @@
+        """
+        # TODO Can't do this can't have the __str__ method be dependent on an execution
+        #  state of the instance
+        #  _samples_analysis may not be populated yet
```
`_samples_analysis may not be populated yet`:

Which is actually an indication that sampling has not occurred! Unsure why `__str__` method should not be dependent on an execution state of the instance, it's purpose is to return a string representation of the instance, so it is natural that as a function of instance life cycle, this representation changes.

davisadam10 replied:
Totally get the need to see the status of the object

my point was I currently get a stack trace when I try and print this object out when the sampling has not been run

Do we want that? 

__str__ should always give you something usable.

This feels dirty to me to have it stack trace.

Id vote for at least handling it and simply saying No Samples or something

davisadam10 replied:
Ok so ive updated this so that it doesnt error also added this to the project settings and also so that we retain the standard python object info for debugging

`<aces.idt.generators.prosumer_camera.IDTGeneratorProsumerCamera object at 0x334358710>

Baseline Exposure : 0
Samples Analysis  : None
LUT Decoding      : None
M                 : None
RGB_w             : None
k                 : None
<aces.idt.framework.project_settings.IDTProjectSettings object at 0x16ac7dc50>

Schema_Version             : 0.1.0
Aces_Transform_Id          : 
Aces_User_Name             : 
Camera_Make                : 
Camera_Model               : 
Iso                        : 800
Temperature                : 6000
Additional_Camera_Settings : 
Lighting_Setup_Description : 
Debayering_Platform        : 
Debayering_Settings        : 
Encoding_Colourspace       : 
Rgb_Display_Colourspace    : sRGB
Cat                        : CAT02
Optimisation_Space         : Oklab
Illuminant_Interpolator    : Linear
Decoding_Method            : Median
Ev_Range                   : [-1.0, 0.0, 1.0]
Grey_Card_Reference        : [0.18, 0.18, 0.18]
Lut_Size                   : 1024
Lut_Smoothing              : 32
Data                       : {'colour_checker': {}, 'grey_card': {}}
Working_Directory          : 
Cleanup                    : False
Reference_Colour_Checker   : ISO 17321-1
Illuminant                 : D60
Sigma                      : 16
File_Type                  : 
Ev_Weights                 : []
Optimization_Kwargs        : {}
`

KelSolaar commented on tests/resources/synthetic_001/synthetic_001.json:
```diff
@@ -7,11 +7,28 @@
     "iso": 100,
     "temperature": 6500,
-    "additional_camera_settings": null,
+    "additional_camera_settings": "",
```
**Major**

I haven't checked but we will want to update the JSON schema accordingly.

davisadam10 replied:
happy either way to be honest as None is also a valid json serialisable, whilst this was proably a copy and paste / auto complete while doing the ProjectSettings i was thinking

string value when empty is "" vs an instance of an object which would be None

Happy to change whichever

KelSolaar reviewed (commented):
I have done a first pass review, lgtm so far! I will need to play with the code a bit to have a better feel for it.

Thank you!

davisadam10 commented on aces/idt/application.py:
```diff
@@ -0,0 +1,311 @@
+        if value not in self.all_generators:
+            raise ValueError(f"Invalid generator name: {value}")
+
+        generator_class = self._all_generators.get(value)
```
changed

davisadam10 commented on aces/idt/application.py:
```diff
@@ -0,0 +1,311 @@
+    def __init__(self):
+        self._project_settings = IDTProjectSettings()
+        self._all_generators = ALL_GENERATORS
+        self._idt_generator = None
```
Yeah agreed, we don't even need to store all_generators here.

Renamed to GENERATORS as the name of the factory
generator_names for the name of the property that gives the names
generator the name of the property that holds the idt_generator.

davisadam10 commented on aces/idt/application.py:
```diff
@@ -0,0 +1,311 @@
+        return self._idt_generator
+
+    @idt_generator.setter
+    def idt_generator(self, value):
```
Resolved and also checked the rest of the file

davisadam10 commented on aces/idt/application.py:
```diff
@@ -0,0 +1,311 @@
+        if not isinstance(value, IDTProjectSettings):
+            raise TypeError("project_settings must be of type IDTProjectSettings")
+
+        # The application holds a single project settings object
+        # When a new project is set the properties are set rather than replace the
+        # object so this can be referenced throughout
+        for name, prop in value.properties:
+            value2 = prop.getter(value)
+            setattr(self._project_settings, name, value2)
```
Yes thats cleaner, i agreed this logic should live in the project settings not the application, application

davisadam10 commented on aces/idt/application.py:
```diff
@@ -0,0 +1,311 @@
+
+
+class IDTGeneratorApplication:
+    """The main application class which handles project loading and saving, the
```
For all docstrings? assuming your meaning use multiline style for multiline and single line for single line style.

Is this something the pre commits can handle?

KelSolaar commented:
I did a first varnish pass, so far so good, I noted though that the tests do not pass if they have not been run once. I haven't looked as to why, a file is missing but generated subsequently.

Notable changes:

- I removed the `sigma` parameter and replaced it with `lut_smoothing`.
- `IDTProjectSettings.from_folder` generate a file which I'm not sure was intended but it seems a bit unexpected.
 
That made me realise that there is a mix of `folder` and `directory` names. We should pick one. I tend to prefer `directory` because we more often about *directory structure* rather than *folder structure*. `DataFolderStructure` --> `DirectoryStructure`?

davisadam10 commented:
DirectoryStructure lets go with that, and well replace all mentions of Folder - I agree

IDTProjectSettings.from_folder -> IDTProjectSettings.from_directory

So the idea is to be able to trawl the directory structure and find the image files for each of the exposure stops/values and then saves the project settings alongside.

Aka i recorded my stuff heres a folder with it in, generate me a project with everything filled in and i can just edit this. (In the world you may not be using the UI).

The longest part and most error prone is filling in the json file for the recordings, so being able to say here to me is helpful.

But perhaps the function is too monolithic and should be seperate functions, where this call creates the IDTProjectSettings from the directory, but then the user has to use the other API to deal with the actual saving of that to disk if they want, rather than doing both in the same func.

davisadam10 commented:
Addressed the remaining notes, I renamed anything with mention of from_folder to use directory and updated the constants too. 

Also made it so the IDTProjectSettings.from_directory now returns the instance of the new project settings and does not also do the saving, this was dirty.

The only outstanding thing I can see is the comment about the functions the getting of the illuminent and the reference checker.

Then need a hand getting the UI back up and running, then we should look at what logic lives in the UI which should move in to the IDTApplication.

Then my suggestion would be to get this back on main, then we branch again and start adding in the new Generators for the pre linearized data, and the cameras which are tonemapped

KelSolaar commented:
Ok, so the UI is back in business, I did a few changes:

- Renamed a few things so that they are a bit more generic and shorter:
  - `IDTMetaData`  -> `Metadata`
  - `IDTMetadataProperty` -> `MetadataProperty`
  - `idt_metadata_property` -> `metadata_property`
  - `BaseSerializable` -> `MixinSerializableProperties` (It is really a mixin)
- `IDTProjectSettings` can now take arguments, i.e., values for the settings.
- `IDTGeneratorApplication` can now take arguments, i.e, generator name and project settings.

I will do another cleanup pass a bit later.

davisadam10 commented:
@KelSolaar tested my end and UI is all working again.

Also had a look over all of your other renames and adjustments, all of which make total sense too.

KelSolaar commented:
Allright, let's merge this one then!
